File: simple-flac-implementation/simple-encode-wav-to-flac.py
# 
# Simple FLAC encoder (Python)
# 
# Copyright (c) 2017 Project Nayuki
# https://www.nayuki.io/page/simple-flac-implementation
# 
# (MIT License)
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
# - The above copyright notice and this permission notice shall be included in
#   all copies or substantial portions of the Software.
# - The Software is provided "as is", without warranty of any kind, express or
#   implied, including but not limited to the warranties of merchantability,
#   fitness for a particular purpose and noninfringement. In no event shall the
#   authors or copyright holders be liable for any claim, damages or other
#   liability, whether in an action of contract, tort or otherwise, arising from,
#   out of or in connection with the Software or the use or other dealings in the
#   Software.
# 

import sys
python3 = sys.version_info.major >= 3


def main(argv):
	if len(argv) != 3:
		sys.exit("Usage: python " + argv[0] + " InFile.wav OutFile.flac")
	with open(argv[1], "rb") as inp:
		with BitOutputStream(open(argv[2], "wb")) as out:
			encode_file(inp, out)


def encode_file(inp, out):
	# Read and parse WAV file headers
	def fail_if(cond, msg):
		if cond:
			raise ValueError(msg)
	fail_if(read_fully(inp, 4) != b"RIFF", "Invalid RIFF file header")
	read_little_int(inp, 4)
	fail_if(read_fully(inp, 4) != b"WAVE", "Invalid WAV file header")
	fail_if(read_fully(inp, 4) != b"fmt ", "Unrecognized WAV file chunk")
	fail_if(read_little_int(inp, 4) != 16, "Unsupported WAV file type")
	fail_if(read_little_int(inp, 2) != 0x0001, "Unsupported WAV file codec")
	numchannels = read_little_int(inp, 2)
	fail_if(not (1 <= numchannels <= 8), "Too many (or few) audio channels")
	samplerate = read_little_int(inp, 4)
	fail_if(not (1 <= samplerate < (1 << 20)), "Sample rate too large or invalid")
	read_little_int(inp, 4)
	read_little_int(inp, 2)
	sampledepth = read_little_int(inp, 2)
	fail_if(sampledepth not in (8,16,24,32), "Unsupported sample depth")
	fail_if(read_fully(inp, 4) != b"data", "Unrecognized WAV file chunk")
	sampledatalen = read_little_int(inp, 4)
	fail_if(sampledatalen <= 0 or sampledatalen % (numchannels * (sampledepth // 8)) != 0, "Invalid length of audio sample data")
	
	# Start writing FLAC file header and stream info metadata block
	out.write_int(32, 0x664C6143)
	out.write_int(1, 1)
	out.write_int(7, 0)
	out.write_int(24, 34)
	out.write_int(16, BLOCK_SIZE - 1)
	out.write_int(16, BLOCK_SIZE - 1)
	out.write_int(24, 0)
	out.write_int(24, 0)
	out.write_int(20, samplerate)
	out.write_int(3, numchannels - 1)
	out.write_int(5, sampledepth - 1)
	numsamples = sampledatalen // (numchannels * (sampledepth // 8))
	out.write_int(36, numsamples)
	for _ in range(16):
		out.write_int(8, 0)
	
	# Read raw samples and encode FLAC audio frames
	i = 0
	while numsamples > 0:
		blocksize = min(numsamples, BLOCK_SIZE)
		encode_frame(inp, i, numchannels, sampledepth, samplerate, blocksize, out)
		numsamples -= blocksize
		i += 1

BLOCK_SIZE = 4096


def read_fully(inp, n):
	result = inp.read(n)
	if len(result) < n:
		raise EOFError()
	return result


def read_little_int(inp, n):
	result = 0
	for (i, b) in enumerate(read_fully(inp, n)):
		result |= (b if python3 else ord(b)) << (i * 8)
	return result


def encode_frame(inp, frameindex, numchannels, sampledepth, samplerate, blocksize, out):
	bytespersample = sampledepth // 8
	samples = [[] for _ in range(numchannels)]
	for _ in range(blocksize):
		for chansamples in samples:
			val = read_little_int(inp, bytespersample)
			if sampledepth == 8:
				val -= 128
			else:
				val -= (val >> (sampledepth - 1)) << sampledepth
			chansamples.append(val)
	
	out.reset_crcs()
	out.write_int(14, 0x3FFE)
	out.write_int(1, 0)
	out.write_int(1, 0)
	out.write_int(4, 7)
	out.write_int(4, (14 if samplerate % 10 == 0 else 13))
	out.write_int(4, numchannels - 1)
	out.write_int(3, {8:1, 16:4, 24:6, 32:0}[sampledepth])
	out.write_int(1, 0)
	out.write_int(8, 0xFC | (frameindex >> 30))
	for i in range(24, -1, -6):
		out.write_int(8, 0x80 | ((frameindex >> i) & 0x3F))
	out.write_int(16, blocksize - 1)
	out.write_int(16, samplerate // (10 if samplerate % 10 == 0 else 1))
	out.write_int(8, out.crc8)
	
	for chansamples in samples:
		encode_subframe(chansamples, sampledepth, out)
	out.align_to_byte()
	out.write_int(16, out.crc16)


def encode_subframe(samples, sampledepth, out):
	out.write_int(1, 0)
	out.write_int(6, 1)  # Verbatim coding
	out.write_int(1, 0)
	for x in samples:
		out.write_int(sampledepth, x)



class BitOutputStream(object):
	
	def __init__(self, out):
		self.out = out
		self.bitbuffer = 0
		self.bitbufferlen = 0
		self.reset_crcs()
	
	
	def reset_crcs(self):
		self.crc8 = 0
		self.crc16 = 0
	
	
	def align_to_byte(self):
		self.write_int((-self.bitbufferlen) % 8, 0)
	
	
	def write_int(self, n, val):
		self.bitbuffer <<= n
		self.bitbuffer |= val & ((1 << n) - 1)
		self.bitbufferlen += n
		while self.bitbufferlen >= 8:
			self.bitbufferlen -= 8
			b = (self.bitbuffer >> self.bitbufferlen) & 0xFF
			self.out.write(bytes((b,)) if python3 else chr(b))
			self.crc8 = CRC8_TABLE[self.crc8 ^ b]
			self.crc16 = CRC16_TABLE[(self.crc16 >> 8) ^ b] ^ ((self.crc16 & 0xFF) << 8)
		self.bitbuffer &= (1 << self.bitbufferlen) - 1
	
	
	def close(self):
		self.out.close()
	
	
	def __enter__(self):
		return self
	
	
	def __exit__(self, type, value, traceback):
		self.close()


CRC8_TABLE = []
CRC16_TABLE = []
for i in range(256):
	temp8 = i
	temp16 = i << 8
	for _ in range(8):
		temp8 = (temp8 << 1) ^ ((temp8 >> 7) * 0x107)
		temp16 = (temp16 << 1) ^ ((temp16 >> 15) * 0x18005)
	CRC8_TABLE.append(temp8)
	CRC16_TABLE.append(temp16)



if __name__ == "__main__":
	main(sys.argv)

File: simple-flac-implementation/SimpleEncodeWavToFlac.java
/* 
 * Simple FLAC encoder (Java)
 * 
 * Copyright (c) 2017 Project Nayuki. (MIT License)
 * https://www.nayuki.io/page/simple-flac-implementation
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * - The above copyright notice and this permission notice shall be included in
 *   all copies or substantial portions of the Software.
 * - The Software is provided "as is", without warranty of any kind, express or
 *   implied, including but not limited to the warranties of merchantability,
 *   fitness for a particular purpose and noninfringement. In no event shall the
 *   authors or copyright holders be liable for any claim, damages or other
 *   liability, whether in an action of contract, tort or otherwise, arising from,
 *   out of or in connection with the Software or the use or other dealings in the
 *   Software.
 */

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.util.zip.DataFormatException;


public final class SimpleEncodeWavToFlac {
	
	public static void main(String[] args) throws IOException, DataFormatException {
		if (args.length != 2) {
			System.err.println("Usage: java SimpleEncodeWavToFlac InFile.wav OutFile.flac");
			System.exit(1);
			return;
		}
		try (InputStream in = new BufferedInputStream(new FileInputStream(args[0]))) {
			try (BitOutputStream out = new BitOutputStream(new BufferedOutputStream(new FileOutputStream(args[1])))) {
				encodeFile(in, out);
			}
		}
	}
	
	
	public static void encodeFile(InputStream in, BitOutputStream out) throws IOException, DataFormatException {
		// Read and parse WAV file headers
		if (!readString(in, 4).equals("RIFF"))
			throw new DataFormatException("Invalid RIFF file header");
		readLittleInt(in, 4);
		if (!readString(in, 4).equals("WAVE"))
			throw new DataFormatException("Invalid WAV file header");
		
		if (!readString(in, 4).equals("fmt "))
			throw new DataFormatException("Unrecognized WAV file chunk");
		if (readLittleInt(in, 4) != 16)
			throw new DataFormatException("Unsupported WAV file type");
		if (readLittleInt(in, 2) != 0x0001)
			throw new DataFormatException("Unsupported WAV file codec");
		int numChannels = readLittleInt(in, 2);
		if (numChannels < 0 || numChannels > 8)
			throw new RuntimeException("Too many (or few) audio channels");
		int sampleRate = readLittleInt(in, 4);
		if (sampleRate <= 0 || sampleRate >= (1 << 20))
			throw new RuntimeException("Sample rate too large or invalid");
		readLittleInt(in, 4);
		readLittleInt(in, 2);
		int sampleDepth = readLittleInt(in, 2);
		if (sampleDepth == 0 || sampleDepth > 32 || sampleDepth % 8 != 0)
			throw new RuntimeException("Unsupported sample depth");
		
		if (!readString(in, 4).equals("data"))
			throw new DataFormatException("Unrecognized WAV file chunk");
		int sampleDataLen = readLittleInt(in, 4);
		if (sampleDataLen <= 0 || sampleDataLen % (numChannels * (sampleDepth / 8)) != 0)
			throw new DataFormatException("Invalid length of audio sample data");
		
		// Start writing FLAC file header and stream info metadata block
		out.writeInt(32, 0x664C6143);
		out.writeInt(1, 1);
		out.writeInt(7, 0);
		out.writeInt(24, 34);
		out.writeInt(16, BLOCK_SIZE - 1);
		out.writeInt(16, BLOCK_SIZE - 1);
		out.writeInt(24, 0);
		out.writeInt(24, 0);
		out.writeInt(20, sampleRate);
		out.writeInt(3, numChannels - 1);
		out.writeInt(5, sampleDepth - 1);
		int numSamples = sampleDataLen / (numChannels * (sampleDepth / 8));
		out.writeInt(18, numSamples >>> 18);
		out.writeInt(18, numSamples >>>  0);
		for (int i = 0; i < 16; i++)
			out.writeInt(8, 0);
		
		// Read raw samples and encode FLAC audio frames
		for (int i = 0; numSamples > 0; i++) {
			int blockSize = Math.min(numSamples, BLOCK_SIZE);
			encodeFrame(in, i, numChannels, sampleDepth, sampleRate, blockSize, out);
			numSamples -= blockSize;
		}
	}
	
	
	private static final int BLOCK_SIZE = 4096;
	
	
	private static String readString(InputStream in, int len) throws IOException {
		byte[] temp = new byte[len];
		for (int i = 0; i < temp.length; i++) {
			int b = in.read();
			if (b == -1)
				throw new EOFException();
			temp[i] = (byte)b;
		}
		return new String(temp, StandardCharsets.UTF_8);
	}
	
	
	private static int readLittleInt(InputStream in, int n) throws IOException {
		int result = 0;
		for (int i = 0; i < n; i++) {
			int b = in.read();
			if (b == -1)
				throw new EOFException();
			result |= b << (i * 8);
		}
		return result;
	}
	
	
	private static void encodeFrame(InputStream in, int frameIndex, int numChannels, int sampleDepth, int sampleRate, int blockSize, BitOutputStream out) throws IOException {
		int[][] samples = new int[numChannels][blockSize];
		int bytesPerSample = sampleDepth / 8;
		for (int i = 0; i < blockSize; i++) {
			for (int ch = 0; ch < numChannels; ch++) {
				int val = 0;
				for (int j = 0; j < bytesPerSample; j++) {
					int b = in.read();
					if (b == -1)
						throw new EOFException();
					val |= b << (j * 8);
				}
				if (sampleDepth == 8)
					samples[ch][i] = val - 128;
				else
					samples[ch][i] = (val << (32 - sampleDepth)) >> (32 - sampleDepth);
			}
		}
		
		out.resetCrcs();
		out.writeInt(14, 0x3FFE);
		out.writeInt(1, 0);
		out.writeInt(1, 0);
		out.writeInt(4, 7);
		out.writeInt(4, sampleRate % 10 == 0 ? 14 : 13);
		out.writeInt(4, numChannels - 1);
		switch (sampleDepth) {
			case  8:  out.writeInt(3, 1);  break;
			case 16:  out.writeInt(3, 4);  break;
			case 24:  out.writeInt(3, 6);  break;
			case 32:  out.writeInt(3, 0);  break;
			default:  throw new IllegalArgumentException();
		}
		out.writeInt(1, 0);
		out.writeInt(8, 0xFC | (frameIndex >>> 30));
		for (int i = 24; i >= 0; i -= 6)
			out.writeInt(8, 0x80 | ((frameIndex >>> i) & 0x3F));
		out.writeInt(16, blockSize - 1);
		out.writeInt(16, sampleRate / (sampleRate % 10 == 0 ? 10 : 1));
		out.writeInt(8, out.crc8);
		
		for (int[] chanSamples : samples)
			encodeSubframe(chanSamples, sampleDepth, out);
		out.alignToByte();
		out.writeInt(16, out.crc16);
	}
	
	
	private static void encodeSubframe(int[] samples, int sampleDepth, BitOutputStream out) throws IOException {
		out.writeInt(1, 0);
		out.writeInt(6, 1);  // Verbatim coding
		out.writeInt(1, 0);
		for (int x : samples)
			out.writeInt(sampleDepth, x);
	}
	
}



final class BitOutputStream implements AutoCloseable {
	
	private OutputStream out;
	private long bitBuffer;
	private int bitBufferLen;
	public int crc8;
	public int crc16;
	
	
	public BitOutputStream(OutputStream out) {
		this.out = out;
		bitBuffer = 0;
		bitBufferLen = 0;
		resetCrcs();
	}
	
	
	public void resetCrcs() {
		crc8 = 0;
		crc16 = 0;
	}
	
	
	public void alignToByte() throws IOException {
		writeInt((64 - bitBufferLen) % 8, 0);
	}
	
	
	public void writeInt(int n, int val) throws IOException {
		bitBuffer = (bitBuffer << n) | (val & ((1L << n) - 1));
		bitBufferLen += n;
		while (bitBufferLen >= 8) {
			bitBufferLen -= 8;
			int b = (int)(bitBuffer >>> bitBufferLen) & 0xFF;
			out.write(b);
			crc8 ^= b;
			crc16 ^= b << 8;
			for (int i = 0; i < 8; i++) {
				crc8 = (crc8 << 1) ^ ((crc8 >>> 7) * 0x107);
				crc16 = (crc16 << 1) ^ ((crc16 >>> 15) * 0x18005);
			}
		}
	}
	
	
	public void close() throws IOException {
		out.close();
	}
	
}
